FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/DicomToDataLakeProcessingJob.cs

```diff
@@ -174,20 +172,42 @@ public async Task<string> ExecuteAsync(IProgress<string> progress, CancellationT
             IProgress<string> progress,
             CancellationToken cancellationToken)
         {
-            var limit = _inputData.EndOffset - _inputData.StartOffset;
-
-            if (limit > 0)
+            long startOffset = _inputData.StartOffset;
+            while (startOffset + SearchChangeFeedLimit <= _inputData.EndOffset)
             {
                 // Get metadata objects from change feeds
-                SearchResult metadataObjectsResult = await GetMetadataObjectsAsync(limit, cancellationToken);
-
-                foreach (JObject metadataObject in metadataObjectsResult.Resources)
-                {
-                    _cacheResult.Resources[DicomConstants.DicomResourceType].Add(metadataObject);
-                }
+                SearchResult metadataObjectsResult = await GetMetadataObjectsAsync(startOffset, SearchChangeFeedLimit, cancellationToken);
 
+                AddDicomMetadataToCache(metadataObjectsResult.Resources);
                 _cacheResult.CacheSize += metadataObjectsResult.ResultSizeInBytes;
+
                 await TryCommitResultAsync(progress, false, cancellationToken);
+                startOffset += SearchChangeFeedLimit;
+            }
+
+            if (startOffset < _inputData.EndOffset)
+            {
+                // Get metadata objects from change feeds
+                SearchResult metadataObjectsResult = await GetMetadataObjectsAsync(startOffset, _inputData.EndOffset - startOffset, cancellationToken);
+
+                AddDicomMetadataToCache(metadataObjectsResult.Resources);
+                _cacheResult.CacheSize += metadataObjectsResult.ResultSizeInBytes;
+            }
+        }
+
+        /// <summary>
+        /// Add the DICOM metadata list extracted from a request to memory cache.
+        /// </summary>
+        private void AddDicomMetadataToCache(List<JObject> dicomMetadataList)
+        {
+            if (!_cacheResult.Resources.ContainsKey(DicomConstants.DicomResourceType))
+            {
+                _cacheResult.Resources[DicomConstants.DicomResourceType] = new List<JObject>();
+            }
+
+            foreach (JObject metadata in dicomMetadataList)
```

Comment:
Maybe we can use AddRange instead of this foreach loop?

Reply:
Updated, thanks.